transport: change interface map to use ifname instead of ifindex

tomereli opened the pull request:
The ieee1905 transport keeps track of network interfaces using a map of
interfaces where the keys are the interface indexes.
Configuration of the transport to allow an interface to take part in the
ieee1905 networking is done by publishing an interface configuration
message on the bus, which is done once in the BTL for all network
interfaces under the bridge.
Some network interfaces indexes may change in the kernel as part of
autoconfiguration which removes the VAPs and brings them back on, which
causes the interfaces not to be tracked by the ieee1905 transport.

The simplest way to overcome this is to keep track of interface names
instead of indexes, as those will not change (and if they will, prplmesh
needs to be restarted).

In addition, fix a bunch of cppcheck issues. While at it, also add a cppcheck shell wrapper.

Fixes #1137 
Fixes #1153

Commit:
transport: use ifname for interfaces configuration

Preparative commit.

The ieee1905 transport keeps track of network interfaces using a map of
interfaces where the keys are the interface indexes.
Configuration of the transport to allow an interface to take part in the
ieee1905 networking is done by publishing an interface configuration
message on the bus, which is done once in the BTL for all network
interfaces under the bridge.
Some network interfaces indexes may change in the kernel as part of
autoconfiguration which removes the VAPs and brings them back on, which
causes the interfaces not to be tracked by the ieee1905 transport.

The simplest way to overcome this is to keep track of interface names
instead of indexes, as those will not change (and if they will, prplmesh
needs to be restarted).

This is the first commit in the series which only changes the
InterfaceConfigurationMessage::Interface to use interface names
instead of index.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
transport: add if_index2name helper function

Preparative commit.
Add helper wrapper function to if_indextoname which returns an
std::string.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
transport: change interface map to use ifname instead of ifindex

The ieee1905 transport keeps track of network interfaces using a map of
interfaces where the keys are the interface indexes.
Configuration of the transport to allow an interface to take part in the
ieee1905 networking is done by publishing an interface configuration
message on the bus, which is done once in the BTL for all network
interfaces under the bridge.
Some network interfaces indexes may change in the kernel as part of
autoconfiguration which removes the VAPs and brings them back on, which
causes the interfaces not to be tracked by the ieee1905 transport.

The simplest way to overcome this is to keep track of interface names
instead of indexes, as those will not change (and if they will, prplmesh
needs to be restarted).

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
transport: fix cppcheck local variable shadow

Fix local variable shadowing issue reported by cppcheck.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

rmelotte commented on framework/transport/ieee1905_transport/ieee1905_transport_local_bus.cpp:
```diff
@@ -156,20 +159,36 @@ bool Ieee1905Transport::send_packet_to_local_bus(Packet &packet)
 
+static void copy_string(char *dst, const char *src, size_t dst_len)
+{
+    const char *src_end = std::find((char *)src, ((char *)src) + dst_len, '\0');
```
Why are you casting away the constness of src? 
(I'm surprised cppcheck doesn't report it)

tomereli replied:
I have no idea - I just copied this from BCL, this is how its implemented there. I will fix it in both places

arnout replied:
Instead of copying it from BCL, move it from BCL to mapf::utils where everyone can use it.

tomereli replied:
Done - wasn't so simple since I had to link BCL to mapf::common and update all callers to use the new API location, but its done now. We need to move everything that should be shared to mapf::common eventually - so that BCL will be merged into mapf::common.

arnout commented on framework/transport/ieee1905_transport/include/mapf/transport/ieee1905_transport_messages.h:
```diff
@@ -268,9 +269,10 @@ class InterfaceConfigurationMessage : public Message {
-            0; // the interface index of the bridge to which this interface belong to (or zero if it does not belong to a bridge)
+        // the linux interface index of the specified interface
+        char ifname[IF_NAMESIZE];
+        // the interface index of the bridge to which this interface belong to (or zero if it does not belong to a bridge)
+        char bridge_ifname[IF_NAMESIZE];
```
As noted in https://github.com/prplfoundation/prplMesh/issues/1137#issuecomment-613941348 I'm not so sure that using ifname for interfaces is a good idea. However, I could live with it just to make progress.

That said, I think it is definitely *not* a good idea to use an interface name for the bridge. If the bridge gets deleted while prplMesh is running, I think we die a thousand deads anyway... On the other hand, if an interface does get renamed, the bridge is the most likely one to be renamed. And finally, for the issue you're trying to fix, there is no need at all for the bridge to be accessed by name rather than by index. So, I tend to prefer not to change it.

That said, if my vague plan goes through, this configuration message will be removed completely anyway, so it doesn't matter much.

arnout commented on common/beerocks/btl/btl_local_bus.cpp:
```diff
@@ -106,30 +107,28 @@ bool transport_socket_thread::configure_ieee1905_transport_interfaces(
-            bridge_if_index;
+        string_utils::copy_string(
+            interface_configuration_request_msg.metadata()->interfaces[n].ifname, iface.c_str(),
+            IF_NAMESIZE);
```
(nitpick) would be nice to add a variant of copy_string that takes a std::string as argument.

tomereli replied:
I don't see the benefit of such API if we're just saving the call to str->c_str().

arnout commented on framework/transport/ieee1905_transport/test/ieee1905_transport_test.cpp:
```diff
@@ -251,18 +251,10 @@ int main(int argc, char *argv[])
-                fprintf(stderr, "unknown interface %s\n", optarg);
-                return false;
-            }
+            ieee1905_bridge_ifname = std::string(optarg, IF_NAMESIZE);
```
This will do a serious out-of-bounds access, because it copies exactly IF_NAMESIZE characters even if optarg is shorter!
```suggestion
            ieee1905_bridge_ifname = std::string(optarg);
```

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport.cpp:
```diff
@@ -74,21 +74,20 @@ void Ieee1905Transport::run()
         for (auto it = network_interfaces_.begin(); it != network_interfaces_.end(); ++it) {
-            unsigned int if_index   = it->first;
-            auto &network_interface = it->second;
+            auto &iface = it->second;
```
I prefer to have such a rename (`network_interface` -> `iface`) in a separate commit.

However, in this case I don't see a reason for the rename. The longer name is more descriptive and its meaning doesn't change, so why rename it?

tomereli replied:
Reverted the change

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_netlink.cpp:
```diff
@@ -70,11 +70,15 @@ int Ieee1905Transport::handle_netlink_message(struct nlmsghdr *msghdr)
         struct ifinfomsg *ifi = (struct ifinfomsg *)NLMSG_DATA(msghdr);
         bool is_active        = (ifi->ifi_flags & IFF_RUNNING) && (ifi->ifi_flags & IFF_UP);
-
+        std::string ifname    = if_index2name(ifi->ifi_index);
```
Note that the message is followed by the attributes of the interface, one of which is IFLA_IFNAME. So we could just parse those attributes instead of opening a socket (which if_index2name does).

(nitpick)

tomereli replied:
I think that parsing the attribute here is less readable than using indextoname, and the overhead of creating a socket is negligible.

arnout commented on framework/transport/ieee1905_transport/include/mapf/transport/ieee1905_transport.h:
```diff
@@ -64,16 +64,20 @@ class Ieee1905Transport {
     struct NetworkInterface {
-        int fd =
-            -1; // the file descriptor of the socket bound to this interface (or -1 if inactive)
+        // the file descriptor of the socket bound to this interface (or -1 if inactive)
```
If you're anyway changing this, please also doxygenify it:
```suggestion
        /// the file descriptor of the socket bound to this interface (or -1 if inactive)
```

tomereli replied:
I had no idea this should be using `///` instead of `//`. Done.

arnout commented on framework/transport/ieee1905_transport/test/ieee1905_transport_test.cpp:
```diff
@@ -388,15 +388,15 @@ int main(int argc, char *argv[])
             using Flags = InterfaceConfigurationRequestMessage::Flags;
             strncpy(interface_configuration_request_msg.metadata()->interfaces[n].ifname,
-                    ieee1905_bridge_ifname.c_str(), IF_NAMESIZE);
+                    ieee1905_bridge_ifname.c_str(), IF_NAMESIZE - 1);
```
This seems to be an unrelated change...

If you're doing this, use copy_string like everywhere else. Because now it's actually still not correct since the terminating '\0' is not added.

tomereli replied:
Changed to use copy_string.

arnout reviewed (approved):
Even though I think it is going in the wrong direction, the right direction is a much bigger change. So we can merge this IMO.

I still have a bunch of review comments which need to be addressed, but I don't need to re-review.

Commit:
framework: utils: add copy_string APIs

Add APIs for safe string copy using std::copy, ported from BCL.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
cmake: bcl: link bcl to mapf::common

Preparative commit.
Since now mapf::common includes utilities which may be used by any
prplmesh component, link the BCL to it so it will be accessible by all
modules which currently use bcl.

In addition, add mapf::common to the CMAKE export set since BCL is exported.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tree-wide: use mapf::utils::copy_string()

Now that we have a single API accessible for all prplmesh components,
use it instead of the ones defined in BCL, BPL and TLVF.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tree-wide: delete non used copy_string implementations

These aren't used anymore, so delete them.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
cppcheck: fix all [noExplicitConstructor] issues

The purpose of the "explicit" keyword is to tell
the compiler that a certain constructor may not be
used to implicitly cast an expression to its class type.

The explicit keyword is an optional decoration for
constructors that take exactly one argument.
It only applies to single-argument constructors since
those are the only constructors that can be used in type casting.

cppcheck marks these as style issues, so fix all of them in the prplmesh
tree.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: add a local shell wrapper for cppcheck

Add a local shell wrapper which runs cppcheck the same way the CI runs
it.

Fixes #1153

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
monitor: fix cppcheck [uninitMemberVar] issues

For example:
 '/builds/prpl-foundation/prplMesh/agent/src/beerocks/monitor/rdkb/monitor_rdkb_hal.h: warning: Member variable 'rdkb_hal_sta_config::snrLowXing' is not initialized in the constructor. [uninitMemberVar]    explicit rdkb_hal_sta_config(const std::string mac_) { mac = mac_; }

While we're at it, fix some other cppcheck [passedByValue] issues by
passing arguments as const reference.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: rdkb task fix cppcheck issue

Fix [useInitializationList] issue by using initialization list to
initialize mac member.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: monitor rssi: Fix [constVariable] cppcheck

Fix to cppcheck style issues constVariable.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: platform: fix cppcheck [virtualCallInConstructor]

Fix cppcheck issue:
 '/builds/prpl-foundation/prplMesh/agent/src/beerocks/slave/platform_manager/platform_manager_thread.h: warning: Virtual function 'on_thread_stop' is called from destructor '~main_thread()' at line 314. Dynamic binding is not used. [virtualCallInConstructor]    virtual void on_thread_stop() override;

By explicitely calling the class on_thread_stop().

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
